kaldb/src/main/java/com/slack/kaldb/server/OpenSearchBulkIngestAPI.java

```diff
@@ -0,0 +1,237 @@
+    }
+  }
+
+  public OpenSearchBulkIngestAPI(
+      DatasetMetadataStore datasetMetadataStore,
+      KaldbConfigs.PreprocessorConfig preprocessorConfig,
+      PrometheusMeterRegistry meterRegistry) {
+    this(datasetMetadataStore, preprocessorConfig, meterRegistry, INITIALIZE_RATE_LIMIT_WARM);
+  }
+
+  public OpenSearchBulkIngestAPI(
+      DatasetMetadataStore datasetMetadataStore,
+      KaldbConfigs.PreprocessorConfig preprocessorConfig,
+      PrometheusMeterRegistry meterRegistry,
+      boolean initializeRateLimitWarm) {
+    this.datasetMetadataStore = datasetMetadataStore;
+    this.preprocessorConfig = preprocessorConfig;
+    this.meterRegistry = meterRegistry;
+    this.rateLimiter =
+        new PreprocessorRateLimiter(
+            meterRegistry,
+            preprocessorConfig.getPreprocessorInstanceCount(),
+            preprocessorConfig.getRateLimiterMaxBurstSeconds(),
+            initializeRateLimitWarm);
+
+    this.configReloadTimer = meterRegistry.timer(CONFIG_RELOAD_TIMER);
+  }
+
+  /**
+   * 1. Kaldb does not support the concept of "updates". It's always an add 2. The "index" is used
+   * as the span name
+   */
+  @Blocking
+  @Post("/_bulk")
+  public HttpResponse addDocument(String bulkRequest) {
+    try {
+      List<IndexRequest> indexRequests = OpenSearchRequest.parseBulkRequest(bulkRequest);
+      Map<String, List<Trace.Span>> docs =
+          OpenSearchRequest.convertIndexRequestToTraceFormat(indexRequests);
+      // our rate limiter doesn't have a way to acquire permits across multiple datasets
+      // so today as a limitation we reject any request that has documents against multiple indexes
+      // We think most indexing requests will be against 1 index
+      if (docs.keySet().size() > 1) {
+        BulkIngestResponse response =
+            new BulkIngestResponse(0, 0, "request must contain only 1 unique index");
+        return HttpResponse.ofJson(INTERNAL_SERVER_ERROR, response);
+      }
+
+      for (Map.Entry<String, List<Trace.Span>> indexDocs : docs.entrySet()) {
+        final String index = indexDocs.getKey();
+        if (!rateLimiterPredicate.test(index, indexDocs.getValue())) {
+          BulkIngestResponse response = new BulkIngestResponse(0, 0, "rate limit exceeded");
+          return HttpResponse.ofJson(TOO_MANY_REQUESTS, response);
+        }
+      }
+      BulkIngestResponse response = produceDocuments(docs);
+      return HttpResponse.ofJson(response);
+    } catch (Exception e) {
+      LOG.error("Request failed ", e);
+      BulkIngestResponse response = new BulkIngestResponse(0, 0, e.getMessage());
+      return HttpResponse.ofJson(INTERNAL_SERVER_ERROR, response);
+    }
+  }
+
+  public BulkIngestResponse produceDocuments(Map<String, List<Trace.Span>> indexDocs) {
+    int totalDocs = indexDocs.values().stream().mapToInt(List::size).sum();
+
+    String transactionId = String.join("_", indexDocs.keySet()) + "_" + System.currentTimeMillis();
```

Comment:
Does this ever get excessively long? Is there a reason we can't use something like a GUID here?

Reply:
The hope is there is always one key in `indexDocs`. We actually today even error out otherwise.

I can definitely change it to GUID, but my initial reason for doing this was this _might_ have enough info if we were debugging Kafka server performance issues. So knowing timestamp and the index that is being used could be valuable. But looking at transaction id might not be even possible.

Happy to change it if you think this is too long